### PR TITLE
Adapter prototype

### PR DESCRIPTION
**Pull request by rolyhudson**
Updated version of #222 
For discussion of #221 
   
### Issues addressed by this PR
Read and write (tables only for now)

### Test files
[Push and Pull Table](https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/Excel_Toolkit/Issues/Issue221-Adapter?csf=1&web=1&e=J6zhkF)

### CODE REVIEW

**Comment by rwemay** [comment]
I had to close the workbooks to make both pull/push work (didn't do much investigation into settings etc.). Seems to work pretty well though, and can see massive potential for this to 'print out' (or read in) pre-formatted data tables. Great stuff.

**Comment by rwemay** [comment]
On styles, I'm not sure if you're using/creating MS styles specifically, but could be one to think about. We've some styles built into some of the templates, and could perhaps pull these for use, and/or have an input that matches them by name - to save a lot of Grasshopper (or other UI) components in creating the styles outside of the Excel interface. We should have some BHoM objects that carry the style eventually (sorry if I've missed these/they exist already!). 

image

**Comment by rolyhudson** [comment]
> I had to close the workbooks to make both pull/push work (didn't do much investigation into settings etc.). Seems to work pretty well though, and can see massive potential for this to 'print out' (or read in) pre-formatted data tables. Great stuff.

Yes bit of a restriction there. When using ClosedXML as the files cannot be in use. But no need for Excel installed.

**Comment by rolyhudson** [comment]
@rwemay On styles:
Nothing really happening here yet with styles. 
Just a glimmer of an idea that maybe everything stylistic can be generalised with a [CSS ruleset syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/Syntax#CSS_rulesets) `selector {declaration, declaration...} ` 
Which could I think be extended to cover web stuff, dataviz, excel...
With pre-templated files that may not be necessary - but the opportunity to apply styles across toolkits using some standard (with great default settings) objects is quite appealing.

**Comment by rwemay** [comment]
We can always look at alternatives for creating a live link, building on sockets with specific components if necessary. XML will give much wider benefits for read/write. Maybe one thing we could do here is provide a user with a mores specific warning to close the file.

**Comment by AndreBalasian** [comment]
Good development @rolyhudson .
ClosedXML lends itself really well for this type of work with spreadsheets.
I look forward to using and enhancing this to generate reports and read data without having to rely on Excel running in the background.

**Review by AndreBalasian — DISMISSED** [review]
Well written code and good implementation of ClosedXML.

The only optional recommendation as a precautionary measure is to add a 1 second time delay after writing to an excel file in order to give Windows more time to finalise the file closing process.

File: Excel_oM/Excel_oM.csproj [hunk 1]
```diff
@@ -35,6 +35,19 @@
       <HintPath>C:\ProgramData\BHoM\Assemblies\BHoM.dll</HintPath>
       <Private>False</Private>
     </Reference>
+    <Reference Include="ClosedXML, Version=0.95.3.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\ClosedXML.0.95.3\lib\net40\ClosedXML.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
+    <Reference Include="Data_oM">
+      <HintPath>..\..\..\..\..\..\..\ProgramData\BHoM\Assemblies\Data_oM.dll</HintPath>
```
Comment (awakeman) [inline Excel_oM/Excel_oM.csproj]: This should be an absolute path.

**Review by awakeman — CHANGES_REQUESTED** [review]

**Comment by pawelbaran** [comment]
@rolyhudson I guess we could close this one as we have https://github.com/BHoM/Excel_Toolkit/pull/320 now?

**Comment by pawelbaran** [comment]
Closing this as https://github.com/BHoM/Excel_Toolkit/pull/320 got merged.

**Pull request closed without merging**